Typedefs of template arguments aren't required anywhere

While investigating #412, I realized that IWYU's current handling of typedefs of template arguments must be broken.

Consider this test case:

```c++
// class.h
class Class {};

// typedef_tpl.cc
#include "class.h"

template<class T>
class Container {
 public:
  // Should not be an iwyu violation for T inside the template...
  typedef T value_type;
};

void Instantiations() {
  // ... but when this typedef is used, surely the underlying type must be
  // required?
  Container<Class>::value_type vt;
}
```

Running IWYU on typedef_tpl.cc gives the following:
```
$ include-what-you-use typedef_tpl.cc

typedef_tpl.cc should add these lines:
class Class;

typedef_tpl.cc should remove these lines:
- #include "class.h"  // lines 1-1

The full include-list for typedef_tpl.cc:
class Class;
---
```

So it suggests that Class should be forward-declared instead of included. When I do that, on the other hand:
```
$ include-what-you-use typedef_tpl.cc
typedef_tpl.cc:14:32: error: variable has incomplete type 'Container<Class>::value_type' (aka 'Class')
  Container<Class>::value_type vt;
                               ^
typedef_tpl.cc:2:7: note: forward declaration of 'Class'
class Class;
      ^

(typedef_tpl.cc has correct #includes/fwd-decls)
```

IWYU produces broken code in this case. I haven't isolated where this happens yet, but I thought I'd write it down while I have it repro:ed.

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -3075,19 +3075,11 @@ class InstantiatedTemplateVisitor
     return TraverseSubstTemplateTypeParmTypeHelper(typeloc.getTypePtr());
   }
 
-  // These do the actual work of finding the types to return.  Our
-  // task is made easier since (at least in theory), every time we
-  // instantiate a template type, the instantiation has type
-  // SubstTemplateTypeParmTypeLoc in the AST tree.
-  bool VisitSubstTemplateTypeParmType(clang::SubstTemplateTypeParmType* type) {
-    if (CanIgnoreCurrentASTNode() || CanIgnoreType(type))
-      return true;
-
-    // Figure out how this type was actually written.  clang always
-    // canonicalizes SubstTemplateTypeParmType, losing typedef info, etc.
-    const Type* actual_type = ResugarType(type);
-    CHECK_(actual_type && "If !CanIgnoreType(), we should be resugar-able");
-
+  // This helper is called on every use of a template argument type in an
+  // instantiated template.  Its goal is to determine whether that use should
+  // constitute a full-use by the template caller, and perform other necessary
+  // bookkeeping.
+  void AnalyzeTemplateTypeParmUse(const Type* type) {
     // TODO(csilvers): whenever we report a type use here, we want to
     // do an iwyu check on this type (to see if sub-types are used).
 
@@ -3098,8 +3090,16 @@ class InstantiatedTemplateVisitor
     // in_forward_declare_context.  I think this will require changing
     // in_forward_declare_context to yes/no/maybe.
     if (current_ast_node()->ParentIsA<NestedNameSpecifier>()) {
-      ReportTypeUse(CurrentLoc(), actual_type);
-      return Base::VisitSubstTemplateTypeParmType(type);
+      ReportTypeUse(CurrentLoc(), type);
+      return;
+    }
+
+    // If the immediate parent node is a typedef, then register the new type as
+    // a new name for the template argument.
+    if (const TypedefNameDecl* typedef_decl =
+            current_ast_node()->GetParentAs<TypedefNameDecl>()) {
+      const Type* typedef_type = typedef_decl->getTypeForDecl();
+      template_argument_aliases_.emplace(typedef_type, type);
     }
 
     // If we're inside a typedef, we don't need our full type info --
@@ -3118,40 +3118,70 @@ class InstantiatedTemplateVisitor
     for (const ASTNode* ast_node = current_ast_node();
          ast_node != caller_ast_node_; ast_node = ast_node->parent()) {
       if (ast_node->IsA<TypedefNameDecl>())
-        return Base::VisitSubstTemplateTypeParmType(type);
+        return;
     }
 
     // sizeof(a reference type) is the same as sizeof(underlying type).
     // We have to handle that specially here, or else we'll say the
     // reference is forward-declarable, below.
     if (current_ast_node()->ParentIsA<UnaryExprOrTypeTraitExpr>() &&
-        isa<ReferenceType>(actual_type)) {
-      const ReferenceType* actual_reftype = cast<ReferenceType>(actual_type);
+        isa<ReferenceType>(type)) {
+      const ReferenceType* actual_reftype = cast<ReferenceType>(type);
       ReportTypeUse(CurrentLoc(),
                     actual_reftype->getPointeeTypeAsWritten().getTypePtr());
-      return Base::VisitSubstTemplateTypeParmType(type);
+      return;
     }
 
     // If we're used in a forward-declare context (MyFunc<T>() { T* t; }),
     // or are ourselves a pointer type (MyFunc<Myclass*>()),
     // we don't need to do anything: we're fine being forward-declared.
     if (current_ast_node()->in_forward_declare_context())
-      return Base::VisitSubstTemplateTypeParmType(type);
+      return;
 
     if (current_ast_node()->ParentIsA<PointerType>() ||
         current_ast_node()->ParentIsA<LValueReferenceType>() ||
-        IsPointerOrReferenceAsWritten(actual_type))
-      return Base::VisitSubstTemplateTypeParmType(type);
+        IsPointerOrReferenceAsWritten(type))
+      return;
 
     // We attribute all uses in an instantiated template to the
     // template's caller.
-    ReportTypeUse(caller_loc(), actual_type);
+    ReportTypeUse(caller_loc(), type);
 
     // Also report all previous explicit instantiations (declarations and
     // definitions) as uses of the caller's location.
-    ReportExplicitInstantiations(actual_type);
+    ReportExplicitInstantiations(type);
+  }
+
+  // Our task is made easier since (at least in theory), every time we
+  // instantiate a template type, the instantiation has type
+  // SubstTemplateTypeParmTypeLoc in the AST tree.
+  bool VisitSubstTemplateTypeParmType(clang::SubstTemplateTypeParmType* type) {
+    if (CanIgnoreCurrentASTNode() || CanIgnoreType(type))
+      return true;
+
+    // Figure out how this type was actually written.  clang always
+    // canonicalizes SubstTemplateTypeParmType, losing typedef info, etc.
+    const Type* actual_type = ResugarType(type);
+    CHECK_(actual_type && "If !CanIgnoreType(), we should be resugar-able");
+
+    AnalyzeTemplateTypeParmUse(actual_type);
 
     return Base::VisitSubstTemplateTypeParmType(type);
+  }
+
+  bool VisitTypedefType(clang::TypedefType* type) {
+    if (CanIgnoreCurrentASTNode())
+      return true;
+
+    // Typedefs of template arguments require special handling to ensure that
+    // we record full-uses of those arguments where appropriate.  Those
+    // typedefs are stored in the template_argument_aliases_ map.
+    if (const Type* template_arg_type =
+            GetOrDefault(template_argument_aliases_, type, nullptr)) {
+      AnalyzeTemplateTypeParmUse(template_arg_type);
+    }
+
+    return Base::VisitTypedefType(type);
   }
 
   bool VisitTemplateSpecializationType(TemplateSpecializationType* type) {
@@ -3214,6 +3244,7 @@ class InstantiatedTemplateVisitor
   void Clear() {
     caller_ast_node_ = nullptr;
     resugar_map_.clear();
+    template_argument_aliases_.clear();
     traversed_decls_.clear();
     nodes_to_ignore_.clear();
     cache_storers_.clear();
@@ -3542,6 +3573,12 @@ class InstantiatedTemplateVisitor
   // value, it's a default template parameter, that the
   // template-caller may or may not be responsible for.
   map<const Type*, const Type*> resugar_map_;
+
+  // When we see a full-use of a template argument we need to assign that full
+  // use to the template-caller.  Sometimes those uses are hidden behind
+  // type aliases (typedefs).  This maps those aliases to the underlying
+  // template arguments.
+  map<const Type*, const Type*> template_argument_aliases_;
 
   // Used to avoid recursion in the *Helper() methods.
   set<const Decl*> traversed_decls_;

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -3075,14 +3075,8 @@ class InstantiatedTemplateVisitor
     return TraverseSubstTemplateTypeParmTypeHelper(typeloc.getTypePtr());
   }
 
-  // This helper is called on every use of a template argument type in an
-  // instantiated template.  Its goal is to determine whether that use should
-  // constitute a full-use by the template caller, and perform other necessary
-  // bookkeeping.
-  void AnalyzeTemplateTypeParmUse(const Type* type) {
-    // TODO(csilvers): whenever we report a type use here, we want to
-    // do an iwyu check on this type (to see if sub-types are used).
-
+  // Check whether a use of a template parameter is a full use.
+  bool IsTemplateTypeParmUseFullUse(const Type* type) {
     const ASTNode* node = MostElaboratedAncestor(current_ast_node());
 
     // If we're a nested-name-specifier class (the Foo in Foo::bar),
@@ -3092,16 +3086,7 @@ class InstantiatedTemplateVisitor
     // in_forward_declare_context.  I think this will require changing
     // in_forward_declare_context to yes/no/maybe.
     if (node->ParentIsA<NestedNameSpecifier>()) {
-      ReportTypeUse(CurrentLoc(), type);
-      return;
-    }
-
-    // If the immediate parent node is a typedef, then register the new type as
-    // a new name for the template argument.
-    if (const TypedefNameDecl* typedef_decl =
-            node->GetParentAs<TypedefNameDecl>()) {
-      const Type* typedef_type = typedef_decl->getTypeForDecl();
-      template_argument_aliases_.emplace(typedef_type, type);
+      return true;
     }
 
     // If we're inside a typedef, we don't need our full type info --
@@ -3120,7 +3105,7 @@ class InstantiatedTemplateVisitor
     for (const ASTNode* ast_node = node; ast_node != caller_ast_node_;
          ast_node = ast_node->parent()) {
       if (ast_node->IsA<TypedefNameDecl>()) {
-        return;
+        return false;
       }
       if (ast_node->IsA<TemplateSpecializationType>()) {
         // If we hit a template specialization node before the typedef then we
@@ -3134,22 +3119,70 @@ class InstantiatedTemplateVisitor
     // reference is forward-declarable, below.
     if (node->ParentIsA<UnaryExprOrTypeTraitExpr>() &&
         isa<ReferenceType>(type)) {
-      const ReferenceType* actual_reftype = cast<ReferenceType>(type);
-      ReportTypeUse(CurrentLoc(),
-                    actual_reftype->getPointeeTypeAsWritten().getTypePtr());
-      return;
+      return true;
     }
 
     // If we're used in a forward-declare context (MyFunc<T>() { T* t; }),
     // or are ourselves a pointer type (MyFunc<Myclass*>()),
     // we don't need to do anything: we're fine being forward-declared.
     if (node->in_forward_declare_context())
-      return;
+      return false;
 
     if (node->ParentIsA<PointerType>() ||
         node->ParentIsA<LValueReferenceType>() ||
         IsPointerOrReferenceAsWritten(type))
+      return false;
+
+    return true;
+  }
+
+  // This helper is called on every use of a template argument type in an
+  // instantiated template.  Its goal is to determine whether that use should
+  // constitute a full-use by the template caller, and perform other necessary
+  // bookkeeping.
+  void AnalyzeTemplateTypeParmUse(const Type* type) {
+    const ASTNode* node = MostElaboratedAncestor(current_ast_node());
+
+    // If the immediate parent node is a typedef, then register the new type as
+    // a new name for the template argument.
+    if (const TypedefNameDecl* typedef_decl =
+            node->GetParentAs<TypedefNameDecl>()) {
+      const Type* typedef_type = typedef_decl->getTypeForDecl();
+      VERRS(6) << "Registering " << PrintableType(typedef_type)
+               << " as an alias for " << PrintableType(type)
+               << " in the context of this instantiation\n";
+      template_argument_aliases_.emplace(typedef_type, type);
+    }
+
+    // Figure out how this type was actually written.  clang always
+    // canonicalizes SubstTemplateTypeParmType, losing typedef info, etc.
+    const Type* actual_type = ResugarType(type);
+    CHECK_(actual_type &&
+           "Type passed to AnalyzeTemplateTypeParmUse should be resugar-able");
+
+    VERRS(6) << "AnalyzeTemplateTypeParmUse: type = " << PrintableType(type)
+             << ", actual_type = " << PrintableType(actual_type) << '\n';
+
+    if (!IsTemplateTypeParmUseFullUse(actual_type)) {
+      // Non-full uses will already have been reported when they were used as
+      // template arguments, so nothing to do here.
       return;
+    }
+
+    if (isa<ReferenceType>(actual_type)) {
+      // If the argument type is a reference type, then we actually care about
+      // the referred-to type.
+      const ReferenceType* actual_reftype = cast<ReferenceType>(actual_type);
+      type = actual_reftype->getPointeeTypeAsWritten().getTypePtr();
+    }
+
+    // At this point we know we are looking at a full-use of type.  However,
+    // there are still two cases.  If this is a template param that was written
+    // by the user, then we report a use of it.  However, it might also be a
+    // parameter of some implementation detail template which just happens top
+    // involve the user's template parameter somehow.  In this case, we need to
+    // recursively explore the instantiation of *that* template to see if the
+    // user's type is full-used therein.
 
     // We attribute all uses in an instantiated template to the
     // template's caller.
@@ -3167,12 +3200,7 @@ class InstantiatedTemplateVisitor
     if (CanIgnoreCurrentASTNode() || CanIgnoreType(type))
       return true;
 
-    // Figure out how this type was actually written.  clang always
-    // canonicalizes SubstTemplateTypeParmType, losing typedef info, etc.
-    const Type* actual_type = ResugarType(type);
-    CHECK_(actual_type && "If !CanIgnoreType(), we should be resugar-able");
-
-    AnalyzeTemplateTypeParmUse(actual_type);
+    AnalyzeTemplateTypeParmUse(type);
 
     return Base::VisitSubstTemplateTypeParmType(type);
   }

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -1248,13 +1248,19 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   using Base::PrintableCurrentLoc;
   using Base::current_ast_node;
 
+  enum class IgnoreKind {
+    ForUse,
+    ForExpansion,
+  };
+
   //------------------------------------------------------------
   // Pure virtual methods that a subclass must implement.
 
   // Returns true if we are not interested in iwyu information for the
   // given type, where the type is *not* the current AST node.
   // TODO(csilvers): check we're calling this consistent with its API.
-  virtual bool CanIgnoreType(const Type* type) const = 0;
+  virtual bool CanIgnoreType(const Type* type,
+                             IgnoreKind = IgnoreKind::ForUse) const = 0;
 
   // Returns true if we are not interested in doing an iwyu check on
   // the given decl, where the decl is *not* the current AST node.
@@ -2897,9 +2903,21 @@ class InstantiatedTemplateVisitor
 
   string GetSymbolAnnotation() const override { return " in tpl"; }
 
-  bool CanIgnoreType(const Type* type) const override {
-    if (!IsTypeInteresting(type) || !IsKnownTemplateParam(type))
+  bool CanIgnoreType(const Type* type, IgnoreKind ignore_kind =
+                                           IgnoreKind::ForUse) const override {
+    if (!IsTypeInteresting(type))
       return true;
+
+    switch (ignore_kind) {
+      case IgnoreKind::ForUse:
+        if (!IsKnownTemplateParam(type))
+          return true;
+        break;
+      case IgnoreKind::ForExpansion:
+        if (!InvolvesKnownTemplateParam(type))
+          return true;
+        break;
+    }
 
     // If we're a default template argument, we should ignore the type
     // if the template author intend-to-provide it, but otherwise we
@@ -2917,7 +2935,6 @@ class InstantiatedTemplateVisitor
     return IsDefaultTemplateParameter(type) && IsProvidedByTemplate(type);
   }
 
-
   bool IsTypeInteresting(const Type* type) const {
     // We only care about types that would have been dependent in the
     // uninstantiated template: that is, SubstTemplateTypeParmType types
@@ -2932,6 +2949,11 @@ class InstantiatedTemplateVisitor
     // the template being instantiated, so we can be ignored.
     type = RemoveSubstTemplateTypeParm(type);
     return ContainsKey(resugar_map_, type);
+  }
+
+  bool InvolvesKnownTemplateParam(const Type* type) const {
+    return InvolvesTypeForWhich(
+        type, [&](const Type* type) { return IsKnownTemplateParam(type); });
   }
 
   // We ignore function calls in nodes_to_ignore_, which were already
@@ -3050,7 +3072,8 @@ class InstantiatedTemplateVisitor
 
   bool TraverseSubstTemplateTypeParmTypeHelper(
       const clang::SubstTemplateTypeParmType* type) {
-    if (CanIgnoreCurrentASTNode() || CanIgnoreType(type))
+    if (CanIgnoreCurrentASTNode() ||
+        CanIgnoreType(type, IgnoreKind::ForExpansion))
       return true;
 
     const Type* actual_type = ResugarType(type);
@@ -3212,7 +3235,8 @@ class InstantiatedTemplateVisitor
   // instantiate a template type, the instantiation has type
   // SubstTemplateTypeParmTypeLoc in the AST tree.
   bool VisitSubstTemplateTypeParmType(clang::SubstTemplateTypeParmType* type) {
-    if (CanIgnoreCurrentASTNode() || CanIgnoreType(type))
+    if (CanIgnoreCurrentASTNode() ||
+        CanIgnoreType(type, IgnoreKind::ForExpansion))
       return true;
 
     AnalyzeTemplateTypeParmUse(type);
@@ -3692,7 +3716,7 @@ class IwyuAstConsumer
   string GetSymbolAnnotation() const override { return ""; }
 
   // We are interested in all types for iwyu checking.
-  bool CanIgnoreType(const Type* type) const override {
+  bool CanIgnoreType(const Type* type, IgnoreKind) const override {
     return type == nullptr;
   }
 

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -2275,6 +2275,8 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   bool VisitUnaryExprOrTypeTraitExpr(clang::UnaryExprOrTypeTraitExpr* expr) {
     if (CanIgnoreCurrentASTNode())  return true;
 
+    current_ast_node()->set_in_forward_declare_context(false);
+
     // Calling sizeof on a reference-to-X is the same as calling it on X.
     // If sizeof() takes a type, this is easy to check.  If sizeof()
     // takes an expr, it's hard to tell -- GetTypeOf(expr) 'sees through'

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -2681,9 +2681,7 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     //  return false;
 
     // Read past elaborations like 'class' keyword or namespaces.
-    while (ast_node->ParentIsA<ElaboratedType>()) {
-      ast_node = ast_node->parent();
-    }
+    ast_node = MostElaboratedAncestor(ast_node);
 
     // Now there are two options: either we have a type or we have a declaration
     // involving a type.
@@ -3083,13 +3081,15 @@ class InstantiatedTemplateVisitor
     // TODO(csilvers): whenever we report a type use here, we want to
     // do an iwyu check on this type (to see if sub-types are used).
 
+    const ASTNode* node = MostElaboratedAncestor(current_ast_node());
+
     // If we're a nested-name-specifier class (the Foo in Foo::bar),
     // we need our full type info no matter what the context (even if
     // we're a pointer, or a template arg, or whatever).
     // TODO(csilvers): consider encoding this logic via
     // in_forward_declare_context.  I think this will require changing
     // in_forward_declare_context to yes/no/maybe.
-    if (current_ast_node()->ParentIsA<NestedNameSpecifier>()) {
+    if (node->ParentIsA<NestedNameSpecifier>()) {
       ReportTypeUse(CurrentLoc(), type);
       return;
     }
@@ -3097,7 +3097,7 @@ class InstantiatedTemplateVisitor
     // If the immediate parent node is a typedef, then register the new type as
     // a new name for the template argument.
     if (const TypedefNameDecl* typedef_decl =
-            current_ast_node()->GetParentAs<TypedefNameDecl>()) {
+            node->GetParentAs<TypedefNameDecl>()) {
       const Type* typedef_type = typedef_decl->getTypeForDecl();
       template_argument_aliases_.emplace(typedef_type, type);
     }
@@ -3115,16 +3115,17 @@ class InstantiatedTemplateVisitor
     // to require it as well.  TODO(csilvers): this doesn't really
     // make any sense.  Who figures out we need the full type if
     // you do 'Foo<MyClass>::value_type m;'?
-    for (const ASTNode* ast_node = current_ast_node();
-         ast_node != caller_ast_node_; ast_node = ast_node->parent()) {
-      if (ast_node->IsA<TypedefNameDecl>())
+    for (const ASTNode* ast_node = node; ast_node != caller_ast_node_;
+         ast_node = ast_node->parent()) {
+      if (ast_node->IsA<TypedefNameDecl>()) {
         return;
+      }
     }
 
     // sizeof(a reference type) is the same as sizeof(underlying type).
     // We have to handle that specially here, or else we'll say the
     // reference is forward-declarable, below.
-    if (current_ast_node()->ParentIsA<UnaryExprOrTypeTraitExpr>() &&
+    if (node->ParentIsA<UnaryExprOrTypeTraitExpr>() &&
         isa<ReferenceType>(type)) {
       const ReferenceType* actual_reftype = cast<ReferenceType>(type);
       ReportTypeUse(CurrentLoc(),
@@ -3135,11 +3136,11 @@ class InstantiatedTemplateVisitor
     // If we're used in a forward-declare context (MyFunc<T>() { T* t; }),
     // or are ourselves a pointer type (MyFunc<Myclass*>()),
     // we don't need to do anything: we're fine being forward-declared.
-    if (current_ast_node()->in_forward_declare_context())
+    if (node->in_forward_declare_context())
       return;
 
-    if (current_ast_node()->ParentIsA<PointerType>() ||
-        current_ast_node()->ParentIsA<LValueReferenceType>() ||
+    if (node->ParentIsA<PointerType>() ||
+        node->ParentIsA<LValueReferenceType>() ||
         IsPointerOrReferenceAsWritten(type))
       return;
 

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -3120,6 +3120,11 @@ class InstantiatedTemplateVisitor
       if (ast_node->IsA<TypedefNameDecl>()) {
         return;
       }
+      if (ast_node->IsA<TemplateSpecializationType>()) {
+        // If we hit a template specialization node before the typedef then we
+        // probably still need a full-use, so stop looking.
+        break;
+      }
     }
 
     // sizeof(a reference type) is the same as sizeof(underlying type).

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -3184,13 +3184,28 @@ class InstantiatedTemplateVisitor
     // recursively explore the instantiation of *that* template to see if the
     // user's type is full-used therein.
 
-    // We attribute all uses in an instantiated template to the
-    // template's caller.
-    ReportTypeUse(caller_loc(), type);
+    if (IsKnownTemplateParam(type)) {
+      // We attribute all uses in an instantiated template to the
+      // template's caller.
+      ReportTypeUse(caller_loc(), type);
 
-    // Also report all previous explicit instantiations (declarations and
-    // definitions) as uses of the caller's location.
-    ReportExplicitInstantiations(type);
+      // Also report all previous explicit instantiations (declarations and
+      // definitions) as uses of the caller's location.
+      ReportExplicitInstantiations(type);
+    } else {
+      const Decl* decl = TypeToDeclAsWritten(type);
+      if (const auto* cts_decl =
+              dyn_cast<ClassTemplateSpecializationDecl>(decl)) {
+        if (ContainsKey(traversed_decls_, decl))
+          return;  // avoid recursion & repetition
+        traversed_decls_.insert(decl);
+
+        VERRS(6)
+            << "Recursively traversing " << PrintableDecl(cts_decl)
+            << " which was full-used and involves a known template param\n";
+        TraverseDecl(const_cast<ClassTemplateSpecializationDecl*>(cts_decl));
+      }
+    }
   }
 
   // Our task is made easier since (at least in theory), every time we

--- a/iwyu_ast_util.cc
+++ b/iwyu_ast_util.cc
@@ -233,6 +233,14 @@ bool IsElaborationNode(const ASTNode* ast_node) {
   return elaborated_type && elaborated_type->getKeyword() != clang::ETK_None;
 }
 
+const ASTNode* MostElaboratedAncestor(const ASTNode* ast_node) {
+  // Read past elaborations like 'class' keyword or namespaces.
+  while (ast_node->ParentIsA<ElaboratedType>()) {
+    ast_node = ast_node->parent();
+  }
+  return ast_node;
+}
+
 bool IsQualifiedNameNode(const ASTNode* ast_node) {
   if (ast_node == nullptr)
     return false;

--- a/iwyu_ast_util.cc
+++ b/iwyu_ast_util.cc
@@ -122,6 +122,7 @@ using llvm::ArrayRef;
 using llvm::PointerUnion;
 using llvm::cast;
 using llvm::dyn_cast;
+using llvm::dyn_cast_or_null;
 using llvm::errs;
 using llvm::isa;
 using llvm::raw_string_ostream;
@@ -1155,6 +1156,26 @@ const Type* RemoveSubstTemplateTypeParm(const Type* type) {
     return type;
 }
 
+bool InvolvesTypeForWhich(const Type* type,
+                          std::function<bool(const Type*)> pred) {
+  type = RemoveSubstTemplateTypeParm(type);
+  if (pred(type))
+    return true;
+  const Decl* decl = TypeToDeclAsWritten(type);
+  if (const auto* cts_decl =
+          dyn_cast_or_null<ClassTemplateSpecializationDecl>(decl)) {
+    const TemplateArgumentList& tpl_args = cts_decl->getTemplateArgs();
+    for (const TemplateArgument& tpl_arg : tpl_args.asArray()) {
+      if (const Type* arg_type = GetTemplateArgAsType(tpl_arg)) {
+        if (InvolvesTypeForWhich(arg_type, pred)) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
 bool IsPointerOrReferenceAsWritten(const Type* type) {
   type = RemoveElaboration(type);
   return isa<PointerType>(type) || isa<LValueReferenceType>(type);
@@ -1282,13 +1303,20 @@ GetTplTypeResugarMapForClassNoComponentTypes(const clang::Type* type) {
   if (!tpl_spec_type)
     return retval;
 
-  // Get the list of template args that apply to the decls.
-  const NamedDecl* decl = TypeToDeclForContent(tpl_spec_type);
-  const ClassTemplateSpecializationDecl* tpl_decl = DynCastFrom(decl);
-  if (!tpl_decl)   // probably because tpl_spec_type is a dependent type
-    return retval;
-  const TemplateArgumentList& tpl_args
-      = tpl_decl->getTemplateInstantiationArgs();
+  // Pull the template arguments out of the specialization type. If this is
+  // a ClassTemplateSpecializationDecl specifically, we want to
+  // get the arguments therefrom to correctly handle default arguments.
+  const TemplateArgument* tpl_args = tpl_spec_type->getArgs();
+  unsigned num_args = tpl_spec_type->getNumArgs();
+
+  const NamedDecl* decl = TypeToDeclAsWritten(tpl_spec_type);
+  const auto* cls_tpl_decl = dyn_cast<ClassTemplateSpecializationDecl>(decl);
+  if (cls_tpl_decl) {
+    const TemplateArgumentList& tpl_arg_list =
+        cls_tpl_decl->getTemplateInstantiationArgs();
+    tpl_args = tpl_arg_list.data();
+    num_args = tpl_arg_list.size();
+  }
 
   // TemplateSpecializationType only includes explicitly specified
   // types in its args list, so we start with that.  Note that an
@@ -1304,12 +1332,13 @@ GetTplTypeResugarMapForClassNoComponentTypes(const clang::Type* type) {
     // (the latter are all desugared).  If there's a match, update
     // the mapping.
     for (const Type* type : arg_components) {
-      for (unsigned i = 0; i < tpl_args.size(); ++i) {
+      for (unsigned i = 0; i < num_args; ++i) {
         if (const Type* arg_type = GetTemplateArgAsType(tpl_args[i])) {
           if (GetCanonicalType(type) == arg_type) {
             retval[arg_type] = type;
             VERRS(6) << "Adding a template-class type of interest: "
-                     << PrintableType(type) << "\n";
+                     << PrintableType(arg_type) << " -> " << PrintableType(type)
+                     << "\n";
             explicit_args.insert(i);
           }
         }
@@ -1318,7 +1347,7 @@ GetTplTypeResugarMapForClassNoComponentTypes(const clang::Type* type) {
   }
 
   // Now take a look at the args that were not filled explicitly.
-  for (unsigned i = 0; i < tpl_args.size(); ++i) {
+  for (unsigned i = 0; i < num_args; ++i) {
     if (ContainsKey(explicit_args, i))
       continue;
     if (const Type* arg_type = GetTemplateArgAsType(tpl_args[i])) {

--- a/iwyu_ast_util.h
+++ b/iwyu_ast_util.h
@@ -376,6 +376,11 @@ class CurrentASTNodeUpdater {
 // uses ElaboratedType for namespaces ('ns::Foo myvar').
 bool IsElaborationNode(const ASTNode* ast_node);
 
+// Walk up to parents of the given node so long as each parent is an
+// elaboration node (in the sense of IsElaborationNode).
+// Can expand from a node representing 'X' to e.g. 'struct X' or 'mylib::X'.
+const ASTNode* MostElaboratedAncestor(const ASTNode* ast_node);
+
 // See if a given ast_node is a qualified name part of an ElaboratedType
 // node (e.g. 'class ns::Foo x', 'class ::Global x' or 'class Outer::Inner x'.)
 bool IsQualifiedNameNode(const ASTNode* ast_node);

--- a/iwyu_ast_util.h
+++ b/iwyu_ast_util.h
@@ -699,6 +699,11 @@ bool IsClassType(const clang::Type* type);
 // However, vector<T> is *not* converted to vector<int>.
 const clang::Type* RemoveSubstTemplateTypeParm(const clang::Type* type);
 
+// Returns true if any type involved (recursively examining template
+// arguments) satisfies the given predicate.
+bool InvolvesTypeForWhich(const clang::Type* type,
+                          std::function<bool(const clang::Type*)> pred);
+
 // Returns true if type is a pointer type (pointer or reference,
 // looking through elaborations like 'class Foo*' (vs 'Foo*'),
 // but *not* following typedefs (which is why we can't just use

--- a/run_iwyu_tests.py
+++ b/run_iwyu_tests.py
@@ -200,6 +200,7 @@ class OneIwyuTest(unittest.TestCase):
       'relative_exported_mapped_include.cc': ['tests/cxx/subdir'],
       'remove_fwd_decl_when_including.cc': ['.'],
       'self_include.cc': ['.'],
+      'sizeof_in_template_arg.cc': ['.'],
       'sizeof_reference.cc': ['.'],
       'specialization_needs_decl.cc': ['.'],
       'system_namespaces.cc': ['.'],

--- a/tests/cxx/alias_template.cc
+++ b/tests/cxx/alias_template.cc
@@ -11,19 +11,19 @@
 
 #include "tests/cxx/direct.h"
 
-template<class T> struct FullUseTemplateArg {
+template<class T> struct FullUseTemplateArgInSizeof {
   char argument[sizeof(T)];
 };
 
 // Test that we go through alias template and handle aliased template
 // specialization.
-template<class T> using Alias = FullUseTemplateArg<T>;
+template<class T> using Alias = FullUseTemplateArgInSizeof<T>;
 // IWYU: IndirectClass needs a declaration
 // IWYU: IndirectClass is...*indirect.h
 Alias<IndirectClass> alias;
 
 // Test following through entire chain of aliases.
-template<class T> using AliasChain1 = FullUseTemplateArg<T>;
+template<class T> using AliasChain1 = FullUseTemplateArgInSizeof<T>;
 template<class T> using AliasChain2 = AliasChain1<T>;
 // IWYU: IndirectClass needs a declaration
 // IWYU: IndirectClass is...*indirect.h
@@ -32,6 +32,18 @@ AliasChain2<IndirectClass> aliasChain;
 // Test the case when aliased type isn't a template specialization.
 template<class T> using Pointer = T*;
 Pointer<int> intPtr;
+
+template <class T>
+struct FullUseTemplateArgAsVar {
+  T t;
+};
+
+// Test the used class being nested deeper in the alias
+template <typename T>
+using AliasNested = FullUseTemplateArgAsVar<FullUseTemplateArgAsVar<T>>;
+// IWYU: IndirectClass needs a declaration
+// IWYU: IndirectClass is...*indirect.h
+AliasNested<IndirectClass> aliasNested;
 
 /**** IWYU_SUMMARY
 

--- a/tests/cxx/alias_template.cc
+++ b/tests/cxx/alias_template.cc
@@ -41,9 +41,16 @@ struct FullUseTemplateArgAsVar {
 // Test the used class being nested deeper in the alias
 template <typename T>
 using AliasNested = FullUseTemplateArgAsVar<FullUseTemplateArgAsVar<T>>;
+
 // IWYU: IndirectClass needs a declaration
 // IWYU: IndirectClass is...*indirect.h
 AliasNested<IndirectClass> aliasNested;
+
+template <typename T>
+using AliasNested2 = FullUseTemplateArgInSizeof<FullUseTemplateArgInSizeof<T>>;
+// IWYU: IndirectClass needs a declaration
+// IWYU: IndirectClass is...*indirect.h
+AliasNested2<IndirectClass> aliasNested2;
 
 /**** IWYU_SUMMARY
 

--- a/tests/cxx/precomputed_tpl_args.cc
+++ b/tests/cxx/precomputed_tpl_args.cc
@@ -61,11 +61,10 @@ std::bitset<5> bitset;
 // for map<T, SpecializationClass>, we should only consider T.
 
 template<typename T> class TemplatedClass {
-  // TODO(csilvers): IWYU: SpecializationClass is...*precomputed_tpl_args-i1.h
-  // TODO(csilvers): IWYU: std::less is...*precomputed_tpl_args-i1.h
+  // IWYU: SpecializationClass is...*precomputed_tpl_args-i1.h
   // IWYU: SpecializationClass needs a declaration
   std::map<SpecializationClass, T> t1;
-  // TODO(csilvers): IWYU: IndirectClass is...*precomputed_tpl_args-i1.h
+  // IWYU: IndirectClass is...*precomputed_tpl_args-i1.h
   // IWYU: IndirectClass needs a declaration
   std::map<T, IndirectClass> t3;
 };

--- a/tests/cxx/sizeof_in_template_arg.cc
+++ b/tests/cxx/sizeof_in_template_arg.cc
@@ -1,0 +1,41 @@
+//===--- sizeof_in_template_arg.cc - test input file for iwyu -------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+#include "tests/cxx/direct.h"
+
+// This verifies that using sizeof(...) means that the argument of sizeof
+// doesn't count as being in a forward-declare context.  In particular, when
+// it's used as a template argument.
+
+template <unsigned long Size>
+struct Storage {
+  char storage[Size];
+};
+
+template <typename T>
+struct Public {
+  Storage<sizeof(T)> storage;
+};
+
+// IWYU: IndirectClass is...*indirect.h
+// IWYU: IndirectClass needs a declaration
+Public<IndirectClass> p;
+
+/**** IWYU_SUMMARY
+
+tests/cxx/sizeof_in_template_arg.cc should add these lines:
+#include "tests/cxx/indirect.h"
+
+tests/cxx/sizeof_in_template_arg.cc should remove these lines:
+- #include "tests/cxx/direct.h"  // lines XX-XX
+
+The full include-list for tests/cxx/sizeof_in_template_arg.cc:
+#include "tests/cxx/indirect.h"  // for IndirectClass
+
+***** IWYU_SUMMARY */

--- a/tests/cxx/typedef_in_template.cc
+++ b/tests/cxx/typedef_in_template.cc
@@ -25,20 +25,27 @@ class Container {
 
 
 void Declarations() {
-  // These do not need the full type for Class because they're template params.
+  // Just using Container does not need the full type for Class because there
+  // are only aliases made, which do not require full-uses.
 
-  // TODO: This is almost certainly wrong, see bug #431
-  // We should not require the full definition of Class for passing it as a
-  // template argument, but we must require it when the typedef it's aliasing
-  // is full-used.
-  // The bug has instructions for how to provoke the error more obviously.
+  // TODO: But currently this is counted as a full-use because Class is used
+  // inside a template specialization (of Pair) within the definition of
+  // Container.
+  // IWYU: Class is...*typedef_in_template-i1.h
+  // IWYU: Class needs a declaration
+  Container<Class> c;
 
+  // Full-using any of those aliases *should* require a full use of Class.
+
+  // IWYU: Class is...*typedef_in_template-i1.h
   // IWYU: Class needs a declaration
   Container<Class>::value_type vt;
 
+  // IWYU: Class is...*typedef_in_template-i1.h
   // IWYU: Class needs a declaration
   Container<Class>::pair_type pt;
 
+  // IWYU: Class is...*typedef_in_template-i1.h
   // IWYU: Class needs a declaration
   Container<Class>::alias_type at;
 }
@@ -77,6 +84,16 @@ struct IndirectlyUsesAliasedParameter {
 // IWYU: IndirectClass needs a declaration
 IndirectlyUsesAliasedParameter<IndirectClass> b;
 
+template <typename T>
+struct NestedUseOfAliasedParameter {
+  using UserAlias = UsesAliasedParameter<T>;
+  UserAlias a;
+};
+
+// IWYU: IndirectClass is...*indirect.h
+// IWYU: IndirectClass needs a declaration
+NestedUseOfAliasedParameter<IndirectClass> c;
+
 /**** IWYU_SUMMARY
 
 tests/cxx/typedef_in_template.cc should add these lines:
@@ -89,6 +106,6 @@ tests/cxx/typedef_in_template.cc should remove these lines:
 
 The full include-list for tests/cxx/typedef_in_template.cc:
 #include "tests/cxx/indirect.h"  // for IndirectClass
-#include "tests/cxx/typedef_in_template-i1.h"  // for Class (ptr only), Pair
+#include "tests/cxx/typedef_in_template-i1.h"  // for Class, Pair
 
 ***** IWYU_SUMMARY */

--- a/tests/cxx/typedef_in_template.cc
+++ b/tests/cxx/typedef_in_template.cc
@@ -7,6 +7,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "tests/cxx/direct.h"
 #include "tests/cxx/typedef_in_template-d1.h"
 
 template<class T>
@@ -42,16 +43,35 @@ void Declarations() {
   Container<Class>::alias_type at;
 }
 
+// STL containers are often implemented via a complex web of type aliases and
+// helper classes.  Tracking uses through all these layers can be non-trivial.
+// The following are some reduced examples in roughly increasing order of
+// complexity which can serve as helpful test cases while debugging such
+// issues.  They were inspired by libstdc++'s implementation of
+// std::unordered_map, but don't directly correspond to it.
+
+template <typename T>
+struct UsesAliasedParameter {
+  using TAlias = T;
+  TAlias t;
+};
+
+// IWYU: IndirectClass is...*indirect.h
+// IWYU: IndirectClass needs a declaration
+UsesAliasedParameter<IndirectClass> a;
 
 /**** IWYU_SUMMARY
 
 tests/cxx/typedef_in_template.cc should add these lines:
+#include "tests/cxx/indirect.h"
 #include "tests/cxx/typedef_in_template-i1.h"
 
 tests/cxx/typedef_in_template.cc should remove these lines:
+- #include "tests/cxx/direct.h"  // lines XX-XX
 - #include "tests/cxx/typedef_in_template-d1.h"  // lines XX-XX
 
 The full include-list for tests/cxx/typedef_in_template.cc:
+#include "tests/cxx/indirect.h"  // for IndirectClass
 #include "tests/cxx/typedef_in_template-i1.h"  // for Class (ptr only), Pair
 
 ***** IWYU_SUMMARY */

--- a/tests/cxx/typedef_in_template.cc
+++ b/tests/cxx/typedef_in_template.cc
@@ -50,6 +50,8 @@ void Declarations() {
 // issues.  They were inspired by libstdc++'s implementation of
 // std::unordered_map, but don't directly correspond to it.
 
+// Verify that a full-use of an alias of a template parameter is treated as a
+// full-use of that parameter.
 template <typename T>
 struct UsesAliasedParameter {
   using TAlias = T;
@@ -59,6 +61,21 @@ struct UsesAliasedParameter {
 // IWYU: IndirectClass is...*indirect.h
 // IWYU: IndirectClass needs a declaration
 UsesAliasedParameter<IndirectClass> a;
+
+// IWYU: IndirectClass is...*indirect.h
+// IWYU: IndirectClass needs a declaration
+UsesAliasedParameter<IndirectClass>::TAlias a2;
+
+// Try a more complex example, through an additional layer of indirection.
+template <typename T>
+struct IndirectlyUsesAliasedParameter {
+  using TAlias = typename UsesAliasedParameter<T>::TAlias;
+  TAlias t;
+};
+
+// IWYU: IndirectClass is...*indirect.h
+// IWYU: IndirectClass needs a declaration
+IndirectlyUsesAliasedParameter<IndirectClass> b;
 
 /**** IWYU_SUMMARY
 

